core/src/com/unciv/logic/automation/WorkerAutomation.kt

```diff
@@ -362,11 +367,16 @@ class WorkerAutomation(
             tile.terrainFeatures.contains(Constants.marsh) -> "Remove Marsh"
             tile.terrainFeatures.contains(Constants.jungle) -> Constants.tradingPost
             tile.terrainFeatures.contains("Oasis") -> return null
-            tile.terrainFeatures.contains(Constants.forest) -> "Lumber mill"
-            tile.isHill() -> "Mine"
-            tile.baseTerrain in listOf(Constants.grassland, Constants.desert, Constants.plains) -> "Farm"
-            tile.isAdjacentToFreshwater -> "Farm"
-            tile.baseTerrain in listOf(Constants.tundra, Constants.snow) -> Constants.tradingPost
+            tile.terrainFeatures.contains(Constants.forest) && tileImprovements.containsKey("Lumber mill") -> "Lumber mill"
+            tile.isHill() && tileImprovements.containsKey("Mine") -> "Mine"
+            tile.baseTerrain in listOf(Constants.grassland, Constants.desert, Constants.plains)
+                    && tileImprovements.containsKey("Farm") -> "Farm"
+            tile.isAdjacentToFreshwater && tileImprovements.containsKey("Farm") -> "Farm"
+            tile.baseTerrain in listOf(Constants.tundra, Constants.snow) && tileImprovements.containsKey(Constants.tradingPost)
+                -> Constants.tradingPost
```

Comment:
Maybe moving the literals to `Contants` and marking them as would-be-nice-to-get-rid-of there? With a nice name that makes it clear they're automation kludges? And: We could avoid the repeated re-allocation of those listOf's while at it?

Question: The love of contains/containsKey - is `in` not bytecode equivalent and more readable even for Map/ArrayList? Ctrl-B suggests the "equivalent" part is true - so it's a matter of taste?

Reply:
They're both equivalent, for 'key in map' is not as intuitive, since a map for me is made of pairs, not of keys
Agree about the single creation of lists :)
Constants would also be something to track globally - every hardcoded improvement name is a another impediment to modding

Reply:
Actually, you know what? No.
I hope to be able to replace all of this with weighing the *yield* of the improvements, making this moot.

Reply:
But that's for step 2 :)

Reply:
The `in` question was meant globally, that's all over the project. Thanks for explaining.
And "moot" - precisely, but transitioning to symbols recognized by compiler and studio indexer as a stepping stone can help catch oversights. We've done it before.